cmd/dagger/version.go

```diff
@@ -27,3 +61,124 @@ func short() string {
 func long() string {
 	return fmt.Sprintf("%s %s/%s", short(), runtime.GOOS, runtime.GOARCH)
 }
+
+func updateAvailable(ctx context.Context) (string, error) {
+	if engine.IsDevVersion(engine.Version) {
+		return "", nil
+	}
+
+	latest, err := latestVersion(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	// Update is available
+	if semver.Compare(engine.Version, latest) < 0 {
+		return latest, nil
+	}
+
+	return latest, nil
+}
+
+func latestVersion(ctx context.Context) (v string, rerr error) {
+	ctx, span := Tracer().Start(ctx, "check for updates")
+	defer telemetry.End(span, func() error { return rerr })
+
+	imageRef := fmt.Sprintf("%s:latest", engine.EngineImageRepo)
+
+	ref, err := name.ParseReference(imageRef)
+	if err != nil {
+		return "", errors.Wrap(err, "parsing image reference")
+	}
+
+	desc, err := remote.Get(ref,
+		remote.WithContext(ctx),
+		// The default auth keychain parses the same docker credentials as used by the buildkit
+		// session attachable.
+		remote.WithAuthFromKeychain(authn.DefaultKeychain),
+		remote.WithUserAgent(enginetel.Labels{}.WithCILabels().WithAnonymousGitLabels(workdir).UserAgent()),
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to resolve image")
+	}
+
+	annotations, err := manifestAnnotations(desc)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get annotations")
+	}
+
+	version, ok := annotations["org.opencontainers.image.version"]
+	if !ok {
+		return "", errors.New("no version found in annotations")
+	}
+
+	return version, nil
+}
+
+func manifestAnnotations(desc *remote.Descriptor) (map[string]string, error) {
+	annotations := make(map[string]string)
+
+	switch desc.MediaType {
+	case types.OCIImageIndex, types.DockerManifestList:
+		// Handle an OCI image index (v1.IndexManifest)
+		var index v1.IndexManifest
+
+		err := json.Unmarshal(desc.Manifest, &index)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling index: %w", err)
+		}
+
+		// Merge annotations at the index (top-level) if they exist
+		if index.Annotations != nil {
+			for key, value := range index.Annotations {
+				annotations[key] = value
+			}
+		}
+
+		for _, manifest := range index.Manifests {
+			if manifest.Annotations != nil {
+				for key, value := range manifest.Annotations {
+					annotations[key] = value
+				}
+			}
+		}
+
+	case types.OCIManifestSchema1, types.DockerManifestSchema2:
+		// Handle a single image manifest (v1.Manifest)
+		var manifest v1.Manifest
+
+		err := json.Unmarshal(desc.Manifest, &manifest)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling manifest: %w", err)
+		}
+
+		// Copy annotations into the map
+		if manifest.Annotations != nil {
+			for key, value := range manifest.Annotations {
+				annotations[key] = value
+			}
+		}
+
+	default:
+		return nil, fmt.Errorf("unsupported media type: %s\n", desc.MediaType)
+	}
+
+	return annotations, nil
+}
+
+func versionNag(latest string) {
```

Comment:
🙏 

That's it? Nothing more complicated than that?

Reply:
Yep - it'll always be appropriate to use `cmd.OutOrStdout()` and/or `cmd.ErrOrStderr()` since it can be dynamically redirected as needed externally:

https://github.com/dagger/dagger/blob/cf4cf5214036efea8c31003fa318b30ad02f6014/cmd/dagger/engine.go#L93-L94

For commands like `dagger version` it'll (I think) just go straight to stderr, and for commands that wire up the TUI, it'll be configured properly via the above two lines. So just always remember to trust the `cmd` out/err.